README.rst
===========
Rest-Client
===========

.. _description:

REST/JSON client based on Requests.


Usage
=====

.. code-block:: python

   from rest_client import RestClient

   client = RestClient('http://jsonplaceholder.typicode.com',
                       auth=('username', 'password'),
                       options={'timeout': 3.0},
                       user_agent='PlaceHolderClient/1.0')

   # PUT /posts/1
   response = client.call('PUT', ('posts', 1))

   # GET /comments?postId=1
   response = client.call('GET', ('comments',), params=dict(postId=1))



CHANGES.rst
Changelog
=========

Development
-----------

* Update for initial Python 3 support.

1.1.1 (2015-09-29)
------------------

* Fix bug with requests 1.x: missing is_redirect method

1.1.0 (2015-08-13)
------------------

* Normalize error log and augment log with error body

1.0.0 (2015-08-10)
------------------

* Initial release


rest_client/tests.py
import base64
import unittest

from nose_parameterized import parameterized
import httpretty
import mock


class TestBase(unittest.TestCase):

    TEST_BASE = 'http://host/base'
    TEST_BODY = 'MyBodyIsACage'

    def assert_header(self, request, name, value):
        self.assertIn(name, request.headers)
        self.assertEqual(request.headers[name], value)


class TestClientInstance(TestBase):

    def test_session_simple(self):
        from rest_client import RestClient
        import requests
        client = RestClient('http://host/base')
        self.assertIsInstance(client.session, requests.Session)
        self.assertEqual(client.default_options, {'allow_redirects': False})

    def test_session_auth(self):
        from rest_client import RestClient
        auth = ('U', 'P')
        client = RestClient('http://host/base', auth=auth)
        self.assertEqual(client.session.auth, auth)

    def test_session_user_agent(self):
        from rest_client import RestClient
        user_agent = 'Poipoi/1.0'
        client = RestClient('http://host/base', user_agent=user_agent)

        self.assertIn('User-Agent', client.session.headers)
        self.assertIn(user_agent, client.session.headers['User-Agent'])

    def test_options(self):
        from rest_client import RestClient
        options = {'Key': 'Value'}
        client = RestClient('http://host/base', options=options)
        self.assertEqual(client.default_options,
                         {'allow_redirects': False, 'Key': 'Value'})


@mock.patch('rest_client.client.log')
class TestErrorLog(TestBase):

    def test_nominal(self, m_log):
        from rest_client.client import errorlog
        errorlog('TYPE', 'ERROR', 'METHOD', 'URL', 'DETAILS')

        m_log.error.assert_called_once_with(
            'RESTClient type=TYPE error=ERROR detail="DETAILS" '
            'req="METHOD URL"')

    def test_status(self, m_log):
        from rest_client.client import errorlog
        errorlog('TYPE', 'ERROR', 'METHOD', 'URL', 'DETAILS', 400)

        m_log.error.assert_called_once_with(
            'RESTClient type=TYPE error=ERROR detail="DETAILS" '
            'req="METHOD URL" status=400')

    def test_body(self, m_log):
        from rest_client.client import errorlog
        errorlog('TYPE', 'ERROR', 'METHOD', 'URL', 'DETAILS', body='BODY')

        m_log.error.assert_called_once_with(
            'RESTClient type=TYPE error=ERROR detail="DETAILS" '
            'req="METHOD URL"\nBODY')


@mock.patch('rest_client.client.errorlog')
class TestError(TestBase):

    def test_exception(self, m_errorlog):
        from rest_client import RestClient, RequestException

        client = RestClient(self.TEST_BASE)
        with mock.patch.object(client, 'session') as m_session:
            m_request = m_session.request
            m_request.side_effect = RequestException()

            with self.assertRaises(RequestException):
                client.call('GET', ())

        m_errorlog.assert_called_once_with(
            'failure', 'RequestException', 'GET', 'http://host/base',
            m_request.side_effect)

    def test_client_error(self, m_errorlog):
        from rest_client import RestClient, HTTPError

        client = RestClient(self.TEST_BASE)
        with mock.patch.object(client, 'session') as m_session:
            m_response = m_session.request.return_value
            m_response.is_redirect = False
            m_response.raise_for_status.side_effect = HTTPError()
            m_response.status_code = 400
            m_response.json.return_value = {'error': 'ERROR', 'message': 'MSG'}

            with self.assertRaises(HTTPError):
                client.call('GET', ())

        m_errorlog.assert_called_once_with(
            'client', 'ERROR', 'GET', 'http://host/base', 'MSG',
            body=m_response.content, status=400)

    def test_server_error(self, m_errorlog):
        from rest_client import RestClient, HTTPError

        client = RestClient(self.TEST_BASE)
        with mock.patch.object(client, 'session') as m_session:
            m_response = m_session.request.return_value
            m_response.is_redirect = False
            m_response.raise_for_status.side_effect = HTTPError()
            m_response.status_code = 500
            m_response.json.return_value = {'error': 'ERROR', 'message': 'MSG'}

            with self.assertRaises(HTTPError):
                client.call('GET', ())

        m_errorlog.assert_called_once_with(
            'server', 'ERROR', 'GET', 'http://host/base', 'MSG',
            body=m_response.content, status=500)

    def test_redirect(self, m_errorlog):
        from rest_client import RestClient

        client = RestClient(self.TEST_BASE)
        with mock.patch.object(client, 'session') as m_session:
            m_response = m_session.request.return_value
            m_response.is_redirect = True
            m_response.status_code = 301
            m_response.json.return_value = {'error': 'ERROR', 'message': 'MSG'}
            m_response.headers = {'location': 'LOCATION'}
            with self.assertRaises(IOError):
                client.call('GET', ())

        m_errorlog.assert_called_once_with(
            'redirect', 'redirect', 'GET', 'http://host/base', 'LOCATION',
            body=m_response.content, status=301)


class TestApi(TestBase):

    def test_segment_type(self):
        from rest_client import RestClient

        client = RestClient(self.TEST_BASE)
        with self.assertRaises(TypeError):
            client.call('GET', 'thisIsNotATupleOrList')


class TestLegacyRequest(TestBase):

    def test_legacy(self):
        from rest_client import RestClient

        client = RestClient(self.TEST_BASE)
        client.requests_legacy = True  # Force detection of requests 1.x

        with mock.patch.object(client, 'session') as m_session:
            m_response = m_session.request.return_value
            m_response.is_redirect = False
            m_response.status_code = 200

            client.call('GET', [], json='[1, 2, 3]')

            m_session.request.assert_called_once_with(
                allow_redirects=False, data='"[1, 2, 3]"',
                headers={'Content-Type': 'application/json'}, method='GET',
                url='http://host/base')


class TestClientFunctional(TestBase):

    TEST_BASE = 'http://host/base'
    TEST_BODY = 'MyBodyIsACage'

    def setUp(self):
        httpretty.enable()
        httpretty.register_uri('GET', self.TEST_BASE, status=200,
                               body=self.TEST_BODY)

    def tearDown(self):
        httpretty.reset()

    def test_request(self):
        from rest_client import RestClient
        RestClient(self.TEST_BASE).call('GET', ())
        req = httpretty.last_request()

        self.assertEqual(req.path, '/base')
        self.assertEqual(req.body, b'')
        self.assertEqual(req.method, 'GET')
        self.assertEqual(req.request_version, 'HTTP/1.1')

    def test_headers(self):
        from rest_client import RestClient
        RestClient(self.TEST_BASE).call('GET', ())
        req = httpretty.last_request()

        self.assert_header(req, 'host', 'host')
        self.assert_header(req, 'connection', 'keep-alive')
        self.assert_header(req, 'accept', 'application/json')
        self.assert_header(req, 'accept-encoding', 'gzip, deflate')

    def test_response(self):
        from rest_client import RestClient
        response = RestClient(self.TEST_BASE).call('GET', ())

        import requests
        self.assertIsInstance(response, requests.Response)
        self.assertEqual(response.text, self.TEST_BODY)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.url, self.TEST_BASE)

    def test_response_redirect(self):
        headers = {'Location': 'http://go-away.com'}
        httpretty.register_uri('GET', self.TEST_BASE, status=301,
                               adding_headers=headers)

        from rest_client import RestClient

        with self.assertRaises(IOError):
            RestClient(self.TEST_BASE).call('GET', ())

    def test_response_client_error(self):
        httpretty.register_uri('GET', self.TEST_BASE, status=400)

        from rest_client import RestClient, HTTPError

        with self.assertRaises(HTTPError):
            RestClient(self.TEST_BASE).call('GET', ())

    def test_response_server_error(self):
        httpretty.register_uri('GET', self.TEST_BASE, status=500)

        from rest_client import RestClient, HTTPError

        with self.assertRaises(HTTPError):
            RestClient(self.TEST_BASE).call('GET', ())

    @parameterized.expand(['HEAD', 'GET', 'POST', 'PUT', 'PATCH', 'DELETE'])
    def test_method(self, method):
        httpretty.register_uri(method, self.TEST_BASE, status=200)

        from rest_client import RestClient
        RestClient(self.TEST_BASE).call(method, ())
        req = httpretty.last_request()

        self.assertEqual(req.method, method)

    def test_auth(self):
        from rest_client import RestClient
        auth = ('user_name', 'password')
        RestClient(self.TEST_BASE, auth=auth).call('GET', ())
        req = httpretty.last_request()

        auth = base64.b64encode(b'user_name:password').decode('utf-8')
        self.assert_header(req, 'authorization', 'Basic %s' % auth)

    @parameterized.expand([
        ('/base', []),
        ('/base/elements', ['elements']),
        ('/base/elements/1', ['elements', 1]),
        ])
    def test_segment(self, path, segments):
        httpretty.register_uri('GET', 'http://host%s' % path, status=200)

        from rest_client import RestClient
        RestClient(self.TEST_BASE).call('GET', segments)
        req = httpretty.last_request()

        self.assertEqual(req.path, path)


rest_client/client.py
"""Thin REST/JSON client based on Requests."""
try:
    from urlparse import urlsplit, urlunsplit
except ImportError:
    from urllib.parse import urlsplit, urlunsplit

import json
import logging
import posixpath

import requests

from requests.exceptions import RequestException, HTTPError

log = logging.getLogger(__name__)


# TODO: decide to support query params in base_url or not. Simplify _url_join.
# TODO: ServiceClient() -> configure from settings and provide call_project()

def is_redirect(response):
    """Backport request.is_redirect() for python-requests 1.x.
    http://www.python-requests.org/en/latest/api/#requests.Response.is_redirect
    """
    is_redirect = ('location' in response.headers and
                   response.status_code in [301, 302, 303, 307, 308])
    return is_redirect


class RestClient(object):

    """Thin REST/JSON client based on Requests."""

    def __init__(self, base_url, auth=None, options=None, user_agent=None,
                 session=None):
        """Create a new RestClient instance.

        Args:

            base_url (str): Base URL used to form the final url of all requests
                sent with this instance of RestClient.

            auth (Optional[tuple|callable]): Enable Basic/Digest/Custom HTTP
                Auth.

            options (Optional[dict]): Default keyword arguments applied to all
                requests sent with this RestClient.

            user_agent (Optional[str]): Set the User-Agent header of all
                requests sent with this instance of RestClient.

            session (Optional[requests.Session]): Set the session that the
                instance will use for all http requests to send.

        Notes:

            - Redirect are considered failure
        """
        self.base_url = base_url
        self.default_options = {'allow_redirects': False}
        if options is not None:
            self.default_options.update(options)

        # Detect requests version 1.X
        self.requests_legacy = requests.__version__[0] == '1'

        self.session = session if session else requests.Session()
        self.session.auth = auth
        if user_agent is not None:
            user_agent += ' requests/%s' % requests.__version__
            self.session.headers.update({'User-Agent': user_agent})

        # Accept only JSON responses
        self.session.headers.update({'Accept': 'application/json'})

    def _url_join(self, base, *args):
        """Join an arbitrary number of URL path segments together."""
        scheme, netloc, path, query, fragment = urlsplit(base)
        path = path if len(path) else '/'
        path = posixpath.join(path, *[str(x) for x in args])
        return urlunsplit([scheme, netloc, path, query, fragment])

    def call(self, method, segments, **kwargs):
        """Initiate a REST request.

        Args:

            method (str): method for the new request.

            segments (tuple|list): segments of the URL.

            **kwargs: Keyword arguments passed directly to requests.

        Original args of Requests v2.7:

            params (dict|bytes): params to be sent in the query string of the
                request.

            data: Dictionary or bytes to send in the body of the request.

            json: json to send in the body of the request.

            headers (dict|MultiDict): HTTP Headers to send with the request.

            cookies (dict|CookieJar): Cookies to send with the request.

            files (dict): Dictionary of ``'filename': file-like-objects``
                for multipart encoding upload.

            auth (tuple|callable): Enable Basic/Digest/Custom HTTP Auth.

            timeout (float|tuple): How long to wait for the server to send data
                before giving up, as a float, or a
                (connect timeout, read timeout) tuple.

            allow_redirects (bool): Follow or ignore the redirect in responses.

            proxies: Dictionary mapping protocol to the URL of the proxy.

            stream (bool): whether to immediately download the response
                content.

            verify: if True, the SSL cert will be verified. A CA_BUNDLE path
                can also be provided.

            cert: if String, path to ssl client cert file (.pem). If Tuple,
                ('cert', 'key') pair.

        Returns:
            An original request.Response object.
        """
        if not (isinstance(segments, tuple) or isinstance(segments, list)):
            raise TypeError("segments argument must be a tuple or a list")

        url = self._url_join(self.base_url, *segments)

        # Requests 1.x didn't support json natively
        if self.requests_legacy and 'json' in kwargs:
            # 1. Serialize JSON payload
            kwargs['data'] = json.dumps(kwargs.pop('json'))
            # 2. Set Content-Type
            headers = kwargs.setdefault('headers', {})
            headers.update({'Content-Type': 'application/json'})

        opts = self.default_options.copy()
        opts.update(kwargs)

        log.debug('RestClient %s %s params=%s', method, url, opts)

        # Perform the HTTP call
        try:
            resp = self.session.request(method=method, url=url, **opts)
        except RequestException as exc:
            error = exc.__class__.__name__
            errorlog('failure', error, method, url, exc)
            raise

        # Treat redirect as a failure
        if is_redirect(resp):
            errorlog('redirect', 'redirect', method, url,
                     resp.headers['location'], status=resp.status_code,
                     body=resp.content)
            raise IOError('Redirect(%s) %s' % (resp.status_code, resp.reason))

        # Let python-requests detect errors
        # We want the original request exception to keep the same API
        try:
            resp.raise_for_status()
        except HTTPError:
            error_type, error, message = error_from_response(resp)
            errorlog(error_type, error, method, url, message,
                     status=resp.status_code, body=resp.content)
            raise

        log.debug('RestClient %s %s got: %s %s',
                  method, url, resp.status_code, resp.content)

        return resp


def errorlog(type, error, method, url, detail, status=None, body=None):
    """Produce standard error log.

    Example:

        RESTClient type=<> error=<> msg="<>" req="METHOD /<>" [status=<>]
        <body>
    """
    line = 'RESTClient type=%s error=%s detail="%s" req="%s %s"' % (
        type, error, detail, method, url)
    if status is not None:
        line += ' status=%s' % status
    if body is not None:
        line += '\n%s' % body
    log.error(line)


def error_from_response(resp):
    """Classify client/server error and extract error payload (failsafe).

    Expected payload:

    {
        "error": "TypeOfError",
        "message": "Details about this error"
    }
    """
    error_type = 'client' if resp.status_code < 500 else 'server'

    try:
        payload = resp.json()
    except:
        payload = {}
    finally:
        error = payload.get("error", "-")
        message = payload.get("message", "-")

    return error_type, error, message
